pages/docs/couple-your-code/couple-your-code-waveform.md

```diff
@@ -0,0 +1,167 @@
+precice.setMeshVertices(meshID, vertexSize, coords, vertexIDs); 
+delete[] coords;
+
+int displID = precice.getDataID("Displacements", meshID); 
+int forceID = precice.getDataID("Forces", meshID); 
+double* forces = new double[vertexSize*dim];
+double* displacements = new double[vertexSize*dim];
+
+double dt;         // solver timestep size
+double precice_dt; // maximum precice timestep size
+
+precice_dt = precice.initialize();
+while (not simulationDone()){ // time loop
+  dt = beginTimeStep(); // e.g. compute adaptive dt 
+  dt = min(precice_dt, dt);
+  // sampling in the middle of the time step
+  precice.readBlockVectorData(displID, vertexSize, vertexIDs, 0.5 * dt, displacements);
+  setDisplacements(displacements); // displacement at the middle of the time step
+  solveTimeStep(dt); // might be using midpoint rule for time-stepping
+  computeForces(forces);
+  precice.writeBlockVectorData(forceID, vertexSize, vertexIDs, forces);
+  precice_dt = precice.advance(dt);
+  endTimeStep(); // e.g. update variables, increment time
+}
+precice.finalize();
+delete[] vertexIDs, forces, displacements;
+turnOffSolver();
+```
+
+### Use waveform iteration with subcycling
+
+Another example shows how to use subcycling with the waveform iteration API. This is an extension of the example from ["Step 5 - Non-matching time step sizes"](couple-your-code-timestep-sizes.html). The changes are similar to the example above, but additionally the behavior of `isReadDataAvailable` slightly changes, because every time step now provides updated read data:
+
+```cpp
+while (not simulationDone()){ // time loop
+  dt = beginTimeStep(); // e.g. compute adaptive dt 
+  dt = min(precice_dt, dt);
+  if (precice.isReadDataAvailable()){ // always returns true with new API
+    // sampling in the middle of the time step
+    precice.readBlockVectorData(displID, vertexSize, vertexIDs, 0.5 * dt, displacements);
+    setDisplacements(displacements); // displacement at the middle of the time step
+  }
+  solveTimeStep(dt); // might be using midpoint rule for time-stepping
+  if (precice.isWriteDataRequired(dt)){ // only true at the end of the time window
+    computeForces(forces);
+    precice.writeBlockVectorData(forceID, vertexSize, vertexIDs, forces);
+  }
+  precice_dt = precice.advance(dt);
+  endTimeStep(); // e.g. update variables, increment time
+}
+```
+
+### Note on the use of `initializeData`
+
+{% include important.html content="`writeBlockVectorData` *after* `initializeData` writes data corresponding to the *end* of the window." %}
+
+In the very first time window of our coupled simulation we have to provide initial data in order to be able to perform linear interpolation. We have to call `writeBlockVectorData` *before* `initializeData`. This data will be used as initial data for the interpolation. If no initial data is provided, only constant interpolation can be applied in the first window, but linear interpolation is still available in later time windows.
```

Comment:
I am again confused by this. Why don't we perform linear interpolation starting at zero if `initializeData` is not called?
Not calling `initializeData` should give the exact same behavior as calling `initializeData` with zero data.

Reply:
We discussed this issue and agreed on using zero initial data. This still has to be implemented in https://github.com/precice/precice/pull/1029.

Reply:
Revised this in 8ba2573eab4a20317a4cc5c247d21438d9f8f284.